test-setup/src/client_setup.rs

```diff
@@ -0,0 +1,185 @@
+        let (rem, greeting) = self.greeting_codec.decode(bytes).unwrap();
+        assert_eq!(rem, &[]);
+        greeting
+    }
+
+    pub fn decode_command<'a>(&self, bytes: &'a [u8]) -> Command<'a> {
+        let (rem, command) = self.command_codec.decode(bytes).unwrap();
+        assert_eq!(rem, &[]);
+        command
+    }
+
+    pub fn decode_response<'a>(&self, bytes: &'a [u8]) -> Response<'a> {
+        let (rem, response) = self.response_codec.decode(bytes).unwrap();
+        assert_eq!(rem, &[]);
+        response
+    }
+
+    pub fn server_sends_greeting<'a>(&mut self, bytes: &'a [u8]) -> Greeting<'a> {
+        let greeting = self.decode_greeting(bytes);
+
+        self.rt.exec(async {
+            self.server.stream.write_all(bytes).await.unwrap();
+        });
+
+        greeting
+    }
+
+    pub fn client_receives_greeting(&mut self, expected_greeting: Greeting) {
+        self.rt.exec(async {
+            match mem::replace(&mut self.client, Client::Dummy) {
+                Client::NotConnected { stream } => {
+                    let stream = AnyStream::new(stream);
+                    let (flow, greeting) =
+                        ClientFlow::receive_greeting(stream, self.client_flow_options)
+                            .await
+                            .unwrap();
+
+                    assert_eq!(expected_greeting, greeting);
+
+                    self.client = Client::Connected { flow };
+                }
+                Client::Connected { .. } => {
+                    panic!("Must be not connected for receiving greeting");
+                }
+                Client::Dummy => unreachable!(),
+            }
+        })
+    }
+
+    pub fn client_sends_command<'a>(&mut self, bytes: &'a [u8]) -> Command<'a> {
+        let command = self.decode_command(bytes);
+
+        self.rt.exec(async {
+            match &mut self.client {
+                Client::NotConnected { .. } => panic!("Must be connect for sending commands"),
+                Client::Connected { flow } => {
+                    let handle = flow.enqueue_command(command.to_static());
+                    let expected_event = ClientFlowEvent::CommandSent {
+                        handle,
+                        command: command.to_static(),
+                    };
+                    let event = flow.progress().await.unwrap();
+                    assert_eq!(expected_event, event);
+                }
+                Client::Dummy => unreachable!(),
+            }
+            self.server.stream.write_all(bytes).await.unwrap();
+        });
+
+        command
+    }
+
+    pub fn server_receives_command(&mut self, expected_command: Command) {
+        let bytes = self.command_codec.encode(&expected_command).dump();
+
+        self.rt.exec(self.server.read_expected_bytes(&bytes));
+    }
+}
+
+enum Client {
+    NotConnected { stream: TcpStream },
+    Connected { flow: ClientFlow },
+    Dummy,
```

Comment:
question: Would it make sense to use an `Option<Client>` and remove `Client::Dummy`?

Reply:
Yes, maybe, this is WIP and I'm not happy with it either.